core/src/transport.rs

```diff
@@ -221,149 +221,252 @@ pub trait Transport {
     }
 }
 
-/// Event produced by [`Transport::Listener`]s.
+/// The ID of a single listener.
+#[derive(Copy, Clone, Debug, PartialEq, Eq, Hash, PartialOrd, Ord)]
+pub struct ListenerId(u64);
+
+impl ListenerId {
+    /// Creates a `ListenerId` from a non-negative integer.
+    pub fn new(id: u64) -> Self {
+        Self(id)
+    }
+}
+
+impl std::ops::Add<u64> for ListenerId {
+    type Output = Self;
+
+    fn add(self, other: u64) -> Self {
+        Self(self.0 + other)
+    }
+}
+
+/// Event produced by [`Transport`]s.
 ///
-/// Transports are expected to produce `Upgrade` events only for
+/// Transports are expected to produce `Incoming` events only for
 /// listen addresses which have previously been announced via
 /// a `NewAddress` event and which have not been invalidated by
 /// an `AddressExpired` event yet.
-#[derive(Clone, Debug, PartialEq)]
-pub enum ListenerEvent<TUpgr, TErr> {
-    /// The transport is listening on a new additional [`Multiaddr`].
-    NewAddress(Multiaddr),
-    /// An upgrade, consisting of the upgrade future, the listener address and the remote address.
-    Upgrade {
-        /// The upgrade.
-        upgrade: TUpgr,
-        /// The local address which produced this upgrade.
+pub enum TransportEvent<T: Transport> {
+    /// A new address is being listened on.
+    NewAddress {
+        /// The listener that is listening on the new address.
+        listener_id: ListenerId,
```

Comment:
_Still need to give this more thought._

What is the use-case where someone needs to differentiate events by listener? In other words, what is the use-case where a user needs to know about the concept of listeners at all?

Don't users just want to be able to:

- Listen on an address X.
- Stop listening on an address X.

Is it wildcard (e.g. `/ip4/0.0.0.0`) addresses that make this difficult?

Tracking this thought as a comment here for now.

Reply:
//CC @rkuhn as you might need it for ipfs-embed.

Reply:
In general, I think it is a good idea to hand out unique identifiers in asynchronous systems so clients can unambiguously associate events with their actions. The actions in our case are `listen_on` and `dial` and the events are returned on `poll`.

> In other words, what is the use-case where a user needs to know about the concept of listeners at all?

I don't think they need to know about the concept of _listeners_ as such. That is an implementation detail to me. Perhaps we should be calling them `ListenRequestId`s?

I think it is risky to assume that we can go through all usecases and decide that for `Transport`s in particular, we can optimise this away but for the sake of an example, one usecase I can think of is:

Re-triggering a `listen_on` request upon failure. If I have previously issued 4 `listen_on` calls and one of them fails, I want to know, which one failed to accurately track that state. If `listen_on` doesn't hand me a unique identifier for each invocation, this is not possible.

Reply:
I guess what you are suggesting is that the address passed in by the user might already be unique? Wildcard addresses are one example that show that they might not be suitable identifiers. I think there might be others too.

Overall, I think uniformity is more important here than optimising individual situations. I really like how we are recently aligning different abstractions in the codebase.

Reply:
ipfs-embed itself doesn’t care about ListenerIds at this point, but it does expose them to its users. The conclusions people should draw from this information are not very clear, though — most people will probably only be interested when an address starts and stops being listened on. What is the consequence of an error in a listener for the addresses this listener is servicing?

Currently ipfs-embed allows user code to start listening on an address, but there is no API to stop listening. Implementing this missing features will require some sort of ListenerId, will it not?

Reply:
I am undecided on this. 

> In general, I think it is a good idea to hand out unique identifiers in asynchronous systems so clients can unambiguously associate events with their actions. The actions in our case are listen_on and dial and the events are returned on poll.

Generally I agree with this. But then we should be consistent and also add an identifier for initiated dials (and include that in all related events).

On the other hand, I also like reducing the number of publicly exposed types where possible, for the sake of simplicity. And in the end users really only care about what addresses are currently active.

> I think it is risky to assume that we can go through all usecases and decide that for Transports in particular

Before https://github.com/libp2p/rust-libp2p/pull/2123 we also did not have a `ListenerId` included listener events. I actually did that PR back then to include it, but iirc that was mostly so we can expose it to our (Stronghold's) users.

> Re-triggering a listen_on request upon failure. If I have previously issued 4 listen_on calls and one of them fails, I want to know, which one failed to accurately track that state. If listen_on doesn't hand me a unique identifier for each invocation, this is not possible.

Upon failure we can include the multiaddr that the user gave as parameter in the event (even if it's a wildcard address). If the user calls listen_on e.g. 2 times with wildcard addresses and one of them fails, would they really care if it is the first or the second one that failed?  
 
But if we decided to remove the `ListenerId`, I think we should also remove the event variants `ListenerClosed` and `ListenerError`. For the former we emit `AddressExpired` for all affected addresses, and the latter is just informative (I would actually be interested if anyone really handles that case/ takes action?), so if anything we could just add a general `TransportError` variant.

---
 As already mentioned, I am really undecided here. Both directions are valid, so I am fine with either as long as we are consistent.

Reply:
> But then we should be consistent and also add an identifier for initiated dials (and include that in all related events).

Yes. I'd be in favor of this because it also allows us to return the `DialFuture` as part of the events. See https://github.com/libp2p/rust-libp2p/pull/2652#issuecomment-1128536102.

Reply:
I would really like to see events being the only source of new connections because it simplifies client code. Not only outside but also within `libp2p-swarm` :)

Whether or not we include IDs is a question of correctness IMO. I'd vote for including them because it is a a - in my opinion - small price to pay for added correctness and allows to serve more use cases.

Here is a sketch of how I envision the API (modulo errors and bike shedding on names etc):

```rust
pub trait Transport {
    type IncomingConnection;
    type OutgoingConnection;
    
    fn dial(&mut self, address: MultiAddress) -> Result<DialId>;
    fn listen_on(&mut self, address: MultiAddress) -> Result<ListenId>;
    
    fn poll(&mut self, cx: &mut Context<'_>) -> Poll<TransportEvent>;
}

pub enum TransportEvent<T: Transport> {
    NewIncomingConnection {
        listen_id: ListenId,
        connection: T::IncomingConnection
    },
    NewOutgoingConnection {
        dial_id: DialId,
        connection: T::OutgoingConnection
    }
}
```

Reply:
Definitely agree that this would be a very clean and unified API. 
However, I think it would add complexity to the swarm rather than simplify it. The way how concurrent dial currently works is that it calls `dial` for all addresses, pushes the obtained dial futures into a list and hands it to a newly spawned task that drives them. This all happens within the `Swarm::dial_with_handler` call.
If we change the transport according to your proposal, this would be more complicated because we don't obtain the futures until we poll the transport, which only happens in `Swarm::poll`. 
We could for example solve this by using a mpsc channel to "send over" the futures on every `TransportEvent::NewOutgoingConnection` from the swarm to the related spawned task. But I don't think that's very clean.
Also on the transport side I think a transport would on `Transport::dial` still initiate the future, but then have to hold it somewhere until `poll` is called.

That being said, I am generally not opposed to your proposal. I very much appreciate that you are pushing for a simple, unified API. I just think that we have to be careful that we are actually making the implementation and integration easier with this. But maybe we can come up with the clean solution that works well with the concurrent dialing? Interested to hear your ideas @thomaseizinger!

Reply:
> We could for example solve this by using a mpsc channel to "send over" the futures on every `TransportEvent::NewOutgoingConnection` from the swarm to the related spawned task. But I don't think that's very clean.

Agreed.

Reply:
> Also on the transport side I think a transport would on `Transport::dial` still initiate the future, but then have to hold it somewhere until `poll` is called.

That is a problem many implementations of `NetworkBehaviour` have to deal with and is typically solved with a `VecDeque` in here. I agree that it is not pretty but I think an unavoidable consequence of such an asynchronous design :/

> However, I think it would add complexity to the swarm rather than simplify it. The way how concurrent dial currently works is that it calls `dial` for all addresses, pushes the obtained dial futures into a list and hands it to a newly spawned task that drives them. This all happens within the `Swarm::dial_with_handler` call.
> If we change the transport according to your proposal, this would be more complicated because we don't obtain the futures until we poll the transport, which only happens in `Swarm::poll`.

Thanks for digging into the details, I haven't looked at those parts yet!

Nothing particularly novel comes to my mind at the moment but some areas worth exploring could be:

- Delay the call to `self.pool.add_outgoing` until we've collected all futures returned from `Transport::poll`
- Implement concurrent dialing through the pool directly by adding the futures from `TransportEvent::NewOutgoingConnection` individually to the pool
- Move the establishing of connections from `Pool` to a wrapper around `Transport` to also encapsulate (and thus simplify) some of the state required for concurrent dialing. The `Pool` when then only deal with established connections and drive those in individual tasks.

To move forward with this, I think we need more reasons other than this being the "cleaner API". I wonder whether the discussion below about having to `Arc` certain inner transports could be resolved by a design along these lines. Although if we want concurrency, sharing a reference to the inner transport may be unavoidable.

Reply:
Sorry for the late reply here. If I understand it correctly @thomaseizinger, we agree that:
- Returning the dial future via a `TransportEvent` does not work well with the `ConcurrentDial` implementation and there is no trivial solution for this
- We don't have any reason for this change other that it being the "cleaner API"

Thus I would stick with returning the future directly in `Transport::dial`. Please correct me if I am missing something.

Reply:
Yep, works for me!